packages/flutter/lib/src/rendering/flex.dart

```diff
@@ -246,7 +247,7 @@ class RenderFlex extends RenderBox with ContainerRenderObjectMixin<RenderBox, Fl
       while (child != null) {
         int flex = _getFlex(child);
         if (flex > 0) {
-          double childMainSize = spacePerFlex * flex;
+          double childMainSize = math.max(0.0, spacePerFlex * flex);
```

Comment:
after looking at this more, I think your original intuition of having the math.max on the line that sets `spacePerFlex` is correct. It matches similar logic elsewhere.